### PR TITLE
os/: Restructure PM Code

### PR DESCRIPTION
**Pull request by gSahitya-samsung**
This commit change the functional structure of pm_suspend, pm_resume, pm_timedsuspend & pm_sleep APIs.
It merges the pm_domain structure into pm_global to make PM States independent of domain.
It also removes the dependency of pm_activity to change state after board wakeup.

### CODE REVIEW

File: os/pm/pm_checkstate.c
Comment (ritesh55555) [inline os/pm/pm_checkstate.c:152]: space between for and (

File: os/pm/pm_checkstate.c
Comment (ritesh55555) [inline os/pm/pm_checkstate.c:154]: suppose state is in idle, then it will stay in idle according to this . But according to new design , we will either stay in NORMAL incase of lcd domain or STANDBY incase of other domain. Please handle that case

**Comment by ritesh55555** [comment]
@gSahitya-samsung please add more meaningful commit and message

File: os/pm/pm_timedstay.c
Comment (ewoodev) [inline os/pm/pm_timedstay.c:86]: Please chenge this map using pid to using domain.

File: os/include/tinyara/pm/pm.h
Comment (edwakuwaku) [inline os/include/tinyara/pm/pm.h:255]:
```
#ifdef CONFIG_LCD
...
#endif
```

File: os/pm/pm_timedstay.c
Comment (edwakuwaku) [inline os/pm/pm_timedstay.c:127]: (uint32_t)domain

File: os/include/tinyara/pm/pm.h
Comment (kishore-sn) [inline os/include/tinyara/pm/pm.h:507]: This line must be part of description.

File: os/include/tinyara/pm/pm.h
Comment (ritesh55555) [inline os/include/tinyara/pm/pm.h:539]: int -> enum pm_domain_e

File: os/pm/pm_activity.c
Comment (ritesh55555) [inline os/pm/pm_activity.c:191]: int -> enum pm_domain_e

File: os/include/tinyara/pm/pm.h
Comment (edwakuwaku) [inline os/include/tinyara/pm/pm.h:264]: May I know the reason for the count gap?

Comment (ewoodev) [reply os/include/tinyara/pm/pm.h:264]: Hello @edwakuwaku 
we will use fixed domain count 32.

This is for scalability.

Comment (ewoodev) [reply os/include/tinyara/pm/pm.h:264]: The PM merely provides the domain to other functions.

the memory and operating time are the same, there is no need to provide the domain by checking whether the config is enabled or disabled

This is the dependence on the user area.

File: os/include/tinyara/pm/pm.h
Comment (ewoodev) [reply os/include/tinyara/pm/pm.h]: The PM merely provides the domain to other functions.

the memory and operating time are the same, there is no need to provide the domain by checking whether the config is enabled or disabled

This is the dependence on the user area.

File: os/arch/arm/src/amebasmart/amebasmart_pmhelpers.c
Comment (ewoodev) [inline os/arch/arm/src/amebasmart/amebasmart_pmhelpers.c:119]: Do we need to call pm_checkstate in here?
can't we call pm_chagestate(PM_STANDBY)?

**Review by ewoodev — COMMENTED** [review]
Could you please remove domain in power_find_dirref in procfs path?

File: os/arch/arm/src/amebasmart/amebasmart_pmhelpers.c
Comment (gSahitya-samsung) [reply os/arch/arm/src/amebasmart/amebasmart_pmhelpers.c:119]: Suppose if we are in PM_NORMAL state and some interrupt will come. If we do pm_changestate(PM_STANDBY), then our  LCD will switch OFF directly without going to LCD_DIM stage.

File: os/pm/pm_sleep.c
Comment (gSahitya-samsung) [reply os/pm/pm_sleep.c]: Done

File: os/arch/arm/src/amebasmart/amebasmart_pmhelpers.c
Comment (edwakuwaku) [reply os/arch/arm/src/amebasmart/amebasmart_pmhelpers.c:119]: I think changing state when back from sleep (ie. either to standby and normal), can be moved to idle thread logic in amebasmart_idle.c. As to previous implementation, different peripheral has change state as to their callback handler or it's own wakeup logic, and it seems not convenient for maintaining.
Thus, I think it might be better if we leave that to be done in the idle thread.

File: os/arch/arm/src/amebasmart/amebasmart_pmhelpers.c
Comment (edwakuwaku) [inline os/arch/arm/src/amebasmart/amebasmart_pmhelpers.c:115]: Since the PM state change is moved to pm_wakehandler, I don't think we need this timer handler anymore. But you don't need to change it first,
I will do a cleanup once this PR is merged.

File: os/include/tinyara/pm/pm.h
Comment (edwakuwaku) [inline os/include/tinyara/pm/pm.h:102]: I have a question, from app side, once the PM_WIFI_DOMAIN lockcount is 0, can we understand it as there will no be wifi operation ongoing as of that period of time?

Comment (gSahitya-samsung) [reply os/include/tinyara/pm/pm.h:102]: Yes, or in other words you can say "Wifi is ready to go into PM_SLEEP". If it wants to perform some task then it must suspend the PM using PMIOC_SUSPEND

File: os/pm/pm_procfs.c
Comment (edwakuwaku) [inline os/pm/pm_procfs.c:349]: In here, you removed the variable declaration above (ie. index). But it was still being used. Did you enable CONFIG_PM_METRICS when you modified this?

Comment (edwakuwaku) [reply os/pm/pm_procfs.c:349]: On the other hand, the code piece seems abit redundant. Since there is only 4 PM states.
```
for (int index= 0; index < CONFIG_PM_STATES; index++) {
                switch (index) {
                // Each case get different mtrics state
                }
                copysize = snprintf(buffer, buflen, " Time in %s : %d\n", g_power_states[index], // Set final argument from result above);
		buflen -= copysize;
		buffer += copysize;
		totalsize += copysize;
}

```
CONFIG_PM_STATES is just a definition to prevent hardcoding, it might not be defined yet

Comment (gSahitya-samsung) [reply os/pm/pm_procfs.c]: Done

File: os/pm/pm_procfs.c
Comment (edwakuwaku) [inline os/pm/pm_procfs.c:351]: you use mtrics.normal always, is the outcome expected?

File: os/pm/pm_checkstate.c
Comment (ewoodev) [inline os/pm/pm_checkstate.c:147]: if IDLE domain or LCD domain were called suspend, pm state should be PM_NORMAL

Comment (gSahitya-samsung) [reply os/pm/pm_checkstate.c:147]: For LCD case, we will create a different PR

File: os/pm/pm_checkstate.c
Comment (edwakuwaku) [inline os/pm/pm_checkstate.c:146]:
```
} else {
...
```

File: os/pm/pm.h
Comment (ritesh55555) [inline os/pm/pm.h:160]: Any use of accum variable after removing pm_activity ? 
becuase accum variable was only increased by pm_activity.

File: os/pm/pm_wakehandler.c
Comment (ritesh55555) [inline os/pm/pm_wakehandler.c:58]: what is the use of ret variable

File: os/pm/pm_update.c
Comment (ritesh55555) [inline os/pm/pm_update.c:238]: is there any use of these comparison of exit and enter thresholds ? becuase as we remove pm_activity() , i dont think accum variable is changed by any other api, so accum value is always 0 now.  So state change is happening either if we do pm_changestate() or if time interval increases from last state threshold count.
So can we remove unecessary comparision of accum value with enter and exit thresholds? @edwakuwaku @ewoodev @gSahitya-samsung

File: os/pm/pm_activity.c
Comment (ewoodev) [inline os/pm/pm_activity.c:107]: Could you change DEBUGASSERT to return ERROR and set_errno()?

If we control with each domain, we can return ERROR when we use api incorrectly.

File: os/pm/pm_timedstay.c
Comment (ritesh55555) [reply os/pm/pm_timedstay.c:86]: @gSahitya-samsung i think this is still not updated . Can you look into it?

File: os/pm/pm_checkstate.c
Comment (ritesh55555) [inline os/pm/pm_checkstate.c:133]: in pm_update , we were able to change the state based on time threshold , but now there is no way to do that . In pm_update , we were able to set threshold count for each state transition . But now its not there

File: os/pm/pm_checkstate.c
Comment (ritesh55555) [inline os/pm/pm_checkstate.c:114]: why is this necessary?

File: os/pm/pm_procfs.c
Comment (edwakuwaku) [reply os/pm/pm_procfs.c:351]: This seems to be reverted, could you please check again?

File: os/pm/pm_activity.c
Comment (ewoodev) [inline os/pm/pm_activity.c:107]: Please use ERROR

File: os/pm/pm_sleep.c
Comment (ewoodev) [inline os/pm/pm_sleep.c:107]: Please use MSEC2TICK()

**Review by sunghan-chang — APPROVED** [review]

**Pull request merged**
